db/hints/manager.cc

```diff
@@ -326,6 +329,21 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
 }
 
 hint_endpoint_manager& manager::get_ep_manager(const endpoint_id& host_id, const gms::inet_address& ip) {
+    // If this is enabled, we can't rely on the information obtained from `_hint_directory_manager`.
+    if (_uses_host_id) {
+        if (auto it = _ep_managers.find(host_id); it != _ep_managers.end()) {
+            return it->second;
+        }
+    } else {
+        if (const auto maybe_mapping = _hint_directory_manager.get_mapping(host_id, ip)) {
+            return _ep_managers.at(maybe_mapping->first);
+        }
+
+        // If there is no mapping in `_hint_directory_manager` corresponding to either `host_id`, or `ip`,
+        // we need to create a new endpoint manager.
+        _hint_directory_manager.insert_mapping(host_id, ip);
+    }
```

Comment:
No, it works like that by design.

The main problem we need to solve is associate endpoint managers (identified by host IDs) with hint directories (identified by IPs before an upgrade). If we were to try to keep a 1-1 correspondence between the two, we'd need to keep renaming hint directories every time a node changes its IP or a replace operation is performed. We could do it, but we'd incur additional complexity (mostly due to the activity of resource manager and space watchdog).

That's why I decided to associate possibly multiple target nodes with the same hint directory. A simple scenario is like this:
* node A has IP I
* A changes its IP to I'
* node B joins the cluster with IP I
Then, we will save hints towards both node A and node B in the same hint directory because of the mapping kept in `_hint_directory_manager`. Replacing a node is a similar situation except for the fact we get rid of one of the nodes. And here comes the bad side of the story.

If it happens that we associate two different nodes – node A and node B – with the same hint directory and one of those nodes – e.g. node A – leaves the cluster, we will start draining the corresponding endpoint manager – meaning, incoming hints towards node B will be rejected until the draining process finishes, even though we should save those hints. When it's done, we will simply create a new endpoint manager in place of the previous one.

Another problem is we might overly limit the concurrency if we save hints towards multiple nodes via the same endpoint manager.

However, the most important thing is to make sure no data races will occur. And they won't because of the invariant we keep, i.e. at any point in time, there is always exactly one endpoint manager corresponding to a given hint directory.

It's not the smartest way to go about it, but I think it's a low price for making things work until the upgrade process is finished. Once it's done, everything will go back to how it was before this patch, i.e. we regain the 1-to-1 correspondence between endpoint managers and hint directories.